### PR TITLE
Relax id requirement for basic query operations

### PR DESCRIPTION
**Pull request by superkhau**
Connect to #670

### CODE REVIEW

File: test/basic-querying.test.js [hunk 1]
```diff
@@ -620,6 +620,61 @@ describe('basic-querying', function () {
   });
 });
 
+describe('id field', function() {
+  var Todo = getSchema().define('Todo', {
+    id: {type: 'number', id: false},
```
Comment (superkhau) [inline test/basic-querying.test.js]: @raymondfeng Does this look correct? I wasn't sure how to set the `id: false` here.

**Comment by superkhau** [comment]
@raymondfeng Review please. Can you think of any operations I should add or check? I just did the basic CRUD operations.

File: test/basic-querying.test.js [hunk 2]
```diff
@@ -620,6 +620,67 @@ describe('basic-querying', function () {
   });
 });
 
+describe('id field', function() {
+  var Todo;
+
+  before(function setupDb(done) {
+    var db = getSchema();
+    Todo = db.define('Todo', {
+      id: {type: 'number', id: false},
```
Comment (raymondfeng) [inline test/basic-querying.test.js]: It should be `id: false`.

File: test/basic-querying.test.js [hunk 3]
```diff
@@ -620,6 +620,67 @@ describe('basic-querying', function () {
...
+
+
+  it('should not be required for update', function(done) {
+    Todo.update({content: 'Buy ham'}, function(err) {
+      should.not.exist(err);
+      done();
+    });
+  });
+});
+
```
Comment (raymondfeng) [inline test/basic-querying.test.js]: Please also add `count()`.

**Comment by superkhau** [comment]
Raymond Feng [3:23 PM]

```
look for all methods from persisted model
All of the static methods that don’t take an `id` argument should work without PK (edited)
```

File: test/basic-querying.test.js [hunk 4]
```diff
@@ -620,6 +620,73 @@ describe('basic-querying', function () {
   });
 });
 
+describe('id field', function() {
+  var Todo;
+
+  before(function setupDb(done) {
+    var db = getSchema();
+    Todo = db.define('Todo', {
+      id: false,
```
Comment (superkhau) [inline test/basic-querying.test.js]: @raymondfeng Like this?

File: test/basic-querying.test.js [hunk 5]
```diff
@@ -620,6 +620,73 @@ describe('basic-querying', function () {
...
+    Todo.destroyAll();
+    done();
+  });
+
+  it('should not be required for count', function(done) {
+    Todo.count({content: 'Buy eggs'}, function(err) {
+      should.not.exist(err);
+      done();
+    });
+  });
```
Comment (superkhau) [inline test/basic-querying.test.js]: @raymondfeng Added count.

**Comment by superkhau** [comment]
>look for all methods from persisted model

**Static**
- [ ] create
- [ ] upsert, updateOrCreate
- [ ] findOrCreate
- [ ] exists
- [ ] findById
- [ ] find
- [ ] findOne
- [ ] destroyAll, remove, deleteAll
- [ ] updateAll, update
- [ ] destroyById, deleteById, removeById
- [ ] count

- [ ] getIdName
- [ ] setupRemoting
- [ ] diff
- [ ] changes
- [ ] currentCheckpoint
- [ ] replicate
- [ ] createUpdates
- [ ] bulkUpdate
- [ ] getChangeModel
- [ ] getSourceId
- [ ] enableChangeTracking
- [ ] _defineChangeModel
- [ ] recitfyAllChanges
- [ ] handleChangeError
- [ ] rectifyChange
- [ ] findLastChange
- [ ] updateLastChange
- [ ] createChangeStream

**Instance**
- [ ] save
- [ ] isNewRecord
- [ ] destroy, remove, delete
- [ ] updateAttribute(s)
- [ ] reload
- [ ] setId
- [ ] getId
- [ ] getIdName
- [ ]

**Comment by raymondfeng** [comment]
@superkhau Thanks for the list. You should mark the methods that don't require a PK.

Two more things to consider:
- For those methods that require a PK, we should check the model definition and report an error if the model doesn't have a PK
- For those methods that require a PK but the model doesn't have one, we probably should filter them out of REST APIs.

File: test/basic-querying.test.js [hunk 6]
```diff
@@ -620,6 +620,131 @@ describe('basic-querying', function () {
...
+  it('should not be required for findOne', function(done) {
+    Todo.findOne(function(err) {
+      should.not.exist(err);
+      done();
+    });
+  });
+
+  it('should not be required for delete/deleteAll/destroyAll/remove', function(done) {
+    // FIXME: We should add a DAO.delete static method for consistency
+    // (DAO.prototype.delete instance method already exists)
```
Comment (superkhau) [inline test/basic-querying.test.js]: @raymondfeng Thoughts?

Comment (superkhau) [inline test/basic-querying.test.js]: @raymondfeng Thoughts?

**Comment by superkhau** [comment]
@raymondfeng Review please.

@chandadharap I'm bumping this ticket 2 more points as the additional requirements are taking a bit more time to implement.

**tldr**

1. Adding 2 points to this story as scope creep added a bit more effort to the current task
2. Creating a 3 point story to track the second requirement
3. Creating a 1 point doc story to let users know about hiding REST endpoints with ID requirements

---

>For those methods that require a PK, we should check the model definition and report an error if the model doesn't have a PK

I'm creating a new story to track this task, and it will be a 3 pointer.

>For those methods that require a PK but the model doesn't have one, we probably should filter them out of REST APIs.

A third requirement discussed is we will need to add a doc story for @crandmck to let users know that the REST endpoints for functions such as `findById` will be "filtered", meaning we will hide REST endpoints that require an id, but do not have id defined in the model definition

This was all discussed with @raymondfeng over Slack, just giving you the heads up.

**Comment by chandadharap** [comment]
sounds good.

**Comment by superkhau** [comment]
@raymondfeng Actually 3 tests are failing. It's when the code can't access `this.definition.property.id` because it doesn't exist for models accessed through relations. I'm thinking I should validate **both** `this.definition.property.id` and all objects in the `this.definition._id` array for the `id` property since it gets automatically generated there. Thoughts?

**Comment by superkhau** [comment]
@slnode test please

**Comment by superkhau** [comment]
@slnode test please

**Comment by superkhau** [comment]
@slnode test please

**Comment by superkhau** [comment]
@slnode test please

**Comment by superkhau** [comment]
@raymondfeng Ready for review.

File: test/basic-querying.test.js [hunk 7]
```diff
@@ -620,6 +620,172 @@ describe('basic-querying', function () {
   });
 });
 
+describe('queries', function() {
```
Comment (ritch) [inline test/basic-querying.test.js]: This could be a bit more descriptive

File: lib/model-definition.js [hunk 8]
```diff
@@ -302,3 +302,25 @@ ModelDefinition.prototype.toJSON = function (forceRebuild) {
...
+ */
+ModelDefinition.prototype.hasIdDefined = function() {
+  var isDefined = false;
+
+  if (this.properties.id)
+    isDefined = true;
+
+  // check if id is defined by a relation
+  this.ids().forEach(function(item) {
+    if (item.hasOwnProperty('id'))
```
Comment (ritch) [inline lib/model-definition.js]: What if `id` is `false`? I think `id` must be `=== true`.

**Comment by ritch** [comment]
@superkhau - I had a couple minor comments, otherwise LGTM.

File: test/basic-querying.test.js (hunk 7 above)
Comment (superkhau) [inline test/basic-querying.test.js]: When you run the test, it forms the BDD sentence. It is purposely left generic because this test covers a variety of different query types.

File: lib/model-definition.js (hunk 8 above)
Comment (superkhau) [inline lib/model-definition.js]: The check for `false` happens before this. If the model definition has `id: false`, I believe `this.properties.id` is not set at all. Same goes for `this.ids()`.
